File: backend/app/schemas/processing.py
from __future__ import annotations
from datetime import datetime
from typing import Any
from pydantic import Field
from app.models.enums import ProcessingLogLevelEnum, ProcessingStatusEnum
from .common import BaseSchema, IdSchema, CreatedAtSchema
from .report import ReportShortRead
from .template import MlTemplateShortRead
from .upload import ReportUploadShortRead
from .user import UserShortRead

class ProcessingTaskLaunchRequest(BaseSchema):
    report_id: int
    report_upload_id: int
    ml_template_id: int | None = None
    created_by: int | None = None
    priority: int = Field(default=5, ge=1, le=10)
    params_json: dict[str, Any] = Field(default_factory=dict)

class ProcessingTaskUpdate(BaseSchema):
    ml_template_id: int | None = None
    status: ProcessingStatusEnum | None = None
    priority: int | None = Field(None, ge=1, le=10)
    progress: int | None = Field(None, ge=0, le=100)
    params_json: dict[str, Any] | None = None
    quality_score: float | None = Field(None, ge=0, le=1)
    warning_count: int | None = Field(None, ge=0)
    error_count: int | None = Field(None, ge=0)
    retry_count: int | None = Field(None, ge=0)
    error_summary: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None

class ProcessingTaskShortRead(IdSchema):
    report_id: int
    report_upload_id: int
    ml_template_id: int | None = None
    status: ProcessingStatusEnum
    progress: int
    priority: int
    created_at: datetime

class ProcessingTaskRead(CreatedAtSchema, IdSchema):
    report_id: int
    report_upload_id: int
    ml_template_id: int | None = None
    created_by: int | None = None
    celery_task_id: str | None = None
    status: ProcessingStatusEnum
    priority: int
    progress: int
    params_json: dict[str, Any]
    quality_score: float | None = None
    warning_count: int
    error_count: int
    retry_count: int
    error_summary: str | None = None
    queued_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None

class ProcessingLogCreate(BaseSchema):
    processing_task_id: int
    level: ProcessingLogLevelEnum
    stage: str = Field(..., min_length=1, max_length=100)
    message: str = Field(..., min_length=1)
    context_json: dict[str, Any] = Field(default_factory=dict)

class ProcessingLogRead(IdSchema):
    processing_task_id: int
    level: ProcessingLogLevelEnum
    stage: str
    message: str
    context_json: dict[str, Any]
    created_at: datetime

class TaskErrorCreate(BaseSchema):
    processing_task_id: int
    error_code: str = Field(..., min_length=1, max_length=100)
    error_type: str = Field(..., min_length=1, max_length=100)
    field_path: str | None = Field(None, max_length=255)
    row_number: int | None = Field(None, ge=1)
    source_value: str | None = None
    details: str | None = None
    is_critical: bool = False

class TaskErrorRead(IdSchema):
    processing_task_id: int
    error_code: str
    error_type: str
    field_path: str | None = None
    row_number: int | None = None
    source_value: str | None = None
    details: str | None = None
    is_critical: bool
    created_at: datetime

class ProcessingTaskDetailRead(ProcessingTaskRead):
    report: ReportShortRead
    report_upload: ReportUploadShortRead
    ml_template: MlTemplateShortRead | None = None
    creator: UserShortRead | None = None
    logs: list[ProcessingLogRead] = Field(default_factory=list)
    errors: list[TaskErrorRead] = Field(default_factory=list)

File: frontend/src/shared/types/processing.ts
export type ProcessingStatus =
  | 'queued'
  | 'running'
  | 'success'
  | 'failed'
  | 'retry'
  | 'cancelled';

export interface ProcessingTask {
  id: number;
  report_id: number;
  report_upload_id: number;
  ml_template_id: number | null;
  created_by: number | null;
  celery_task_id: string | null;
  status: ProcessingStatus;
  priority: number;
  progress: number;
  params_json: Record<string, unknown>;
  quality_score: number | null;
  warning_count: number;
  error_count: number;
  retry_count: number;
  error_summary: string | null;
  queued_at: string;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
}

export interface ProcessingLog {
  id: number;
  processing_task_id: number;
  level: string;
  stage: string;
  message: string;
  context_json: Record<string, unknown>;
  created_at: string;
}

export interface TaskError {
  id: number;
  processing_task_id: number;
  error_code: string;
  error_type: string;
  field_path: string | null;
  row_number: number | null;
  source_value: string | null;
  details: string | null;
  is_critical: boolean;
  created_at: string;
}

export interface ProcessingTaskDetail extends ProcessingTask {
  report: {
    id: number;
    title: string;
    project_id: number;
    report_type_id: number;
    status: string;
    report_period_start: string;
    report_period_end: string;
    version: number;
    is_archived: boolean;
  };
  report_upload: {
    id: number;
    report_id: number;
    project_id: number;
    report_type_id: number;
    uploaded_by: number;
    original_filename: string;
    upload_version: number;
    is_latest: boolean;
    uploaded_at: string;
  };
  ml_template: {
    id: number;
    code: string;
    name: string;
    template_type: string;
    version: string;
    is_default: boolean;
    is_active: boolean;
    target_report_type_id: number | null;
  } | null;
  creator: {
    id: number;
    email: string;
    full_name: string;
  } | null;
  logs: ProcessingLog[];
  errors: TaskError[];
}

export interface TaskProgressResponse {
  task_id: number;
  status: ProcessingStatus;
  progress: number;
  warning_count: number;
  error_count: number;
  started_at: string | null;
  finished_at: string | null;
  error_summary: string | null;
}